api/server/sdk/watcher.go

```diff
@@ -0,0 +1,264 @@
+	}()
+
+	// wait only as long as context deadline allows
+	select {
+	case err := <-errChan:
+		if err != nil {
+			return status.Errorf(codes.Internal, "error starting watcher: %v", err)
+		} else {
+			return nil
+		}
+	case <-ctx.Done():
+		return status.Error(codes.DeadlineExceeded,
+			"Deadline is reached, server side func exiting")
+	}
+}
+
+func (s *WatcherServer) startVolumeWatcher(ctx context.Context) error {
+	if s.watchConnections == nil {
+		s.watchConnections = make(map[string][]*watchConnection)
+	}
+
+	// wait for driver to be initialized with an non-empty volume watcher
+	for {
+		if s.volumeServer.driver(ctx) == nil {
+			continue
+		}
+
+		volumeChannel, err := s.volumeServer.driver(ctx).GetVolumeWatcher(&api.VolumeLocator{}, make(map[string]string))
+		if err != nil {
+			logrus.Warnf("error getting volume watcher %v", err)
+		}
+		if volumeChannel == nil {
+			continue
+		}
+		s.volumeChannel = volumeChannel
+		goto volumeWatch
+	}
+	// volumeChannel should be waiting for incoming events
+volumeWatch:
+	for {
+		select {
+		case vol := <-s.volumeChannel:
+			s.RLock()
+			for _, client := range s.watchConnections[volumeEventType] {
+				go client.callBack(vol)
+			}
+			s.RUnlock()
+		case <-ctx.Done():
+			logrus.Infof("exiting volume watcher\n")
+			break volumeWatch
+		}
+	}
+
+	return nil
+}
+
+func (w *WatcherServer) volumeWatch(
+	req *api.SdkVolumeWatchRequest,
+	stream api.OpenStorageWatch_WatchServer,
+) error {
+	ctx := stream.Context()
+	if w.volumeServer.cluster() == nil || w.volumeServer.driver(ctx) == nil {
+		return status.Error(codes.Unavailable, "Resource has not been initialized")
+	}
+
+	// if input has deadline, ensure graceful exit within that deadline.
+	deadline, ok := ctx.Deadline()
+	var cancel context.CancelFunc
+	if ok {
+		// create a new context that will get done on deadline
+		ctx, cancel = context.WithTimeout(ctx, time.Until(deadline))
+		defer cancel()
+	}
+
+	client := watchConnection{
+		name:         uuid.New(),
+		eventType:    volumeEventType,
+		eventChannel: make(chan interface{}, eventChannelSize),
+	}
+
+	w.registerWatcher(&client, volumeEventType)
+	defer w.removeWatcher(client.name, volumeEventType)
+
+	group, _ := errgroup.WithContext(ctx)
+	errChan := make(chan error)
+
+	// spawn err-group process.
+	group.Go(func() error {
+		if vols, err := w.volumeServer.driver(ctx).Enumerate(&api.VolumeLocator{}, nil); err != nil {
+			return status.Errorf(
+				codes.Internal,
+				"Failed to enumerate volumes: %v",
+				err.Error())
+		} else {
+			for _, vol := range vols {
+				if !vol.IsPermitted(ctx, api.Ownership_Read) {
+					continue
+				}
+				resp := convertApiVolumeToSdkReponse(vol)
+				if err := stream.Send(resp); err != nil {
+					return err
+				}
+			}
+
+			for event := range client.eventChannel {
+
+				// create a new context that will return error if execution took more than streamTimeout
+				timeoutCtx, timeoutCancelled := context.WithTimeout(ctx, streamTimeout)
```

Comment:
I am not following how this timeoutCtx is being enforced? Isn't steam.Send() a blocking function?